📝 `Developing` Guide page for the `docs` (#343)
neon-mmd/websurfx: docs/developing.md

# Developing

This page of the docs outlines how to get **Websurfx** up and running in a development environment, and outlines the common worflow, different ways to work on the project, a high level overview of how the project works, project structure and the best pratices that should be followed when working on the project.

<details>
  <summary><b>Table of Contents</b></summary>
  <p>

- [Setting up the Development Environment](#setting-up-the-development-environment)

>   - [Local Development](#local-development-)
>   - [Gitpod](#gitpod-)
>   - [NixOS Dev Shell using Nix Flake](#nixos-dev-shell-using-nix-flake-)
>   - [Local Development with Docker Compose](#local-development-with-docker-compose-)

Reviewer: ```suggestion
  - [Local Development](#local-development)
  - [Gitpod](#gitpod)
  - [NixOS Dev Shell using Nix Flake](#nixos-dev-shell-using-nix-flake)
  - [Local Development with Docker Compose](#local-development-with-docker-compose)
```

  - [Project Commands](#project-commands)

>   - [Environmental Variables](#environmental-variables)

Reviewer: ```suggestion
  - [Environment Variables](#environment-variables)
```

- [Git Strategy](#git-strategy)
  - [Flow](#git-flow)
  - [Branches](#git-branch-naming)
  - [Commit emojis](#commit-emojis)
  - [PR Guidelines](#pr-guidelines)
- [Resources for Beginners](#resources-for-beginners)
- [App Info](#app-info)

> - [Code Style Guide](#style-guide)

Reviewer: ```suggestion
- [Code Style Guide](#style-guides)
```

Once you have finished running the above command, Websurfx should now be served on the address http://127.0.0.1:8080. Hot reload is enabled, so making changes to any of the files will trigger the project to being rebuilt.

### Project Commands

#### Basics

- `cargo build`: Builds the project.

> **Note**
> When you build the project first time with the above command it will require the app to compile every dependency in the project which will then be cached on your system. So when you compile the app next time it will only compile for the new changes.

- `cargo run`: Starts the app and serves the project on http://127.0.0.1:8080.

> **Important**
> You must run the build command first.

#### Development

- `cargo watch -q -x "run" -w .`: Starts the development server with hot reloading.
- `cargo fmt -- --check`: Checks the code for proper formatting.
- `cargo clippy`: Lints code to ensure it follows a consistent, neat style.
- `cargo test`: Runs unit tests, integrations tests and doc tests.

### Environment Variables

> All environmental variables are optional. Currently there are not many environmental variables used, as most of the user preferences are stored under the `websurfx` folder (located under the codebase (`websurfx` directory)) in the `config.lua` file.

Reviewer: ```suggestion
Environment variables are optional. Currently there are not many environment variables used, as most of the user preferences are stored under the `websurfx` folder (located under the codebase (`websurfx` directory)) in the `config.lua` file.
```
